fawltydeps/packages.py

```diff
@@ -227,41 +274,32 @@ def resolve_dependencies(
 ) -> Dict[str, Package]:
     """Associate dependencies with corresponding Package objects.
 
-    Use LocalPackageLookup to find Package objects for each of the given
+    Use LocalPackageResolver to find Package objects for each of the given
     dependencies inside the virtualenv given by 'pyenv_path'. When 'pyenv_path'
     is None (the default), look for packages in the current Python environment
     (i.e. equivalent to sys.path).
 
-    For dependencies that cannot be found with LocalPackageLookup,
-    fabricate an identity mapping (a pseudo-package making available an import
-    of the same name as the package, modulo normalization).
-
     Return a dict mapping dependency names to the resolved Package objects.
     """
-    ret = {}
-
-    # consume the iterable once
-    deps = list(dep_names)
+    deps = set(dep_names)  # consume the iterable once
 
+    # This defines the "stack" of resolvers that we will use to convert
+    # dependencies into provided import names. We call .lookup_package() on
+    # each resolver in order until one of them returns a Package object. At
+    # that point we are happy, and don't consult any of the later resolvers.
+    resolvers: List[BasePackageResolver] = [LocalPackageResolver(pyenv_path)]
     if install_deps:
-        logger.info("Installing dependencies into a new temporary Python environment.")
-        if pyenv_path is not None:
-            logger.info(
-                f"Will not use Python environment defined by --pyenv: {pyenv_path}"
-            )
-        local_packages_context = temp_installed_requirements(deps)
-    else:
-        local_packages_context = nullcontext(LocalPackageLookup(pyenv_path))  # type: ignore
-
-    with local_packages_context as local_packages:
-        for name in deps:
-            if name not in ret:
-                package = local_packages.lookup_package(name)
-                if package is None:  # fall back to identity mapping
-                    package = Package.identity_mapping(name)
-                    logger.info(
-                        f"Could not find {name!r} in the current environment. Assuming "
-                        f"it can be imported as {', '.join(sorted(package.import_names))}"
-                    )
-                ret[name] = package
+        resolvers += [TemporaryPipInstallResolver()]
+    resolvers += [IdentityMapping()]
```

Comment:
I would add a TODO-warning here:
```suggestion
    # Identity mapping being at the bottom of the resolvers stack
    # ensures that _all_ deps are matched. If you consider to make 
    # identity mapping optional, remember about handling unresolved dependcies.
    resolvers += [IdentityMapping()]
```

Reply:
```suggestion
    # Identity mapping being at the bottom of the resolvers stack ensures that
    # _all_ deps are matched. TODO: If we make the identity mapping optional,
    # we must remember to properly handle/signal unresolved dependencies.
    resolvers += [IdentityMapping()]
```